src/frolladaptive.c

```diff
@@ -364,3 +378,66 @@ void fadaptiverollsumExact(double *x, uint64_t nx, ans_t *ans, int *k, double fi
     }
   }
 }
+
+/* fast rolling adaptive max - exact
+ * for hasNA=FALSE it will not detect if any NAs were in the input, therefore could produce incorrect result, well documented
+ */
+void frolladaptivemaxExact(double *x, uint64_t nx, ans_t *ans, int *k, double fill, bool narm, int hasna, bool verbose) {
+  if (verbose)
+    snprintf(end(ans->message[0]), 500, _("%s: running in parallel for input length %"PRIu64", hasna %d, narm %d\n"), "frolladaptivemaxExact", (uint64_t)nx, hasna, (int) narm);
+  if (hasna==-1) { // fastest we can get for adaptive max as there is no algo='fast', therefore we drop any NA checks when hasNA=FALSE
+    #pragma omp parallel for num_threads(getDTthreads(nx, true))
+    for (uint64_t i=0; i<nx; i++) {
+      if (i+1 < k[i]) {
+        ans->dbl_v[i] = fill;
+      } else {
+        double w = R_NegInf;
+        for (int j=-k[i]+1; j<=0; j++) { //Rprintf("x[%d+%d] > w: %f > %f: %d\n", i, j, x[i+j], w, x[i+j] > w);
+          if (x[i+j] > w)
+            w = x[i+j];
+        }
+        ans->dbl_v[i] = w;
+      }
+    }
+  } else {
+    if (narm) {
+      #pragma omp parallel for num_threads(getDTthreads(nx, true))
+      for (uint64_t i=0; i<nx; i++) {
+        if (i+1 < k[i]) {
+          ans->dbl_v[i] = fill;
+        } else {
+          int nc = 0;
+          double w = R_NegInf;
+          for (int j=-k[i]+1; j<=0; j++) {
+            if (ISNAN(x[i+j]))
+              nc++;
+            else if (x[i+j] > w)
+              w = x[i+j];
+          }
+          if (nc < k[i])
+            ans->dbl_v[i] = w;
+          else
+            ans->dbl_v[i] = R_NegInf;
+        }
+      }
+    } else {
+      #pragma omp parallel for num_threads(getDTthreads(nx, true))
+      for (uint64_t i=0; i<nx; i++) {
+        if (i+1 < k[i]) {
+          ans->dbl_v[i] = fill;
+        } else {
+          double w = R_NegInf;
+          for (int j=-k[i]+1; j<=0; j++) {
+            if (ISNAN(x[i+j])) {
+              w = NA_REAL;
+              break;
+            } else if (x[i+j] > w) {
+              w = x[i+j];
+            }
+          }
+          ans->dbl_v[i] = w;
+        }
+      }
+    }
+  }
+}
```

Comment:
Can we factor this out to the following or will this interfere with subsequent PRs?
```c
#pragma omp parallel for num_threads(getDTthreads(nx, true))
  for (uint64_t i = 0; i < nx; i++) {
    if (i + 1 < (uint64_t)k[i]) {
      ans->dbl_v[i] = fill;
      continue;
    }

    if (hasna == -1) {
      ans->dbl_v[i] = adaptive_max_nona(x, k[i], i);
    } else if (narm) {
      ans->dbl_v[i] = adaptive_max_narm(x, k[i], i);
    } else {
      ans->dbl_v[i] = adaptive_max_hasna(x, k[i], i);
    }
  }

```

Reply:
As mentioned above regarding tight loops, this change introduces 3 extra branches inside the loop that has to run over `1:length(x)`, we could write it like this but we will trade performance (even if little bit) for readability. Branch prediction, compilers, data are important factor for this, but having performance as a priority always, I do not have to worry about those.

Reply:
@MichaelChirico Not sure if I understand you correctly. The point is to avoid branching inside the loop and calculation between hasna/narm has to happen inside the loop

Reply:
I see, my fault for not reading carefully enough, please ignore :)